src/backend/cdb/cdbmutate.c

```diff
@@ -3781,3 +3787,23 @@ contains_outer_params(Node *node, void *context)
 	}
 	return expression_tree_walker(node, contains_outer_params, context);
 }
+
+static bool
+contain_modifyTable_walker(Node *node, void *context)
+{
+    if (NULL == node)
+        return false;
+
+    if (IsA(node, ModifyTable))
+        return true;
+
+    if (IsA(node, SubPlan))
+    {
+        SubPlanWalkerContext subplan_context;
+        subplan_context.base.node = (Node *) context;
+        subplan_context.bms_subplans = NULL;
+        return plan_tree_walker(node, contain_modifyTable_walker, &subplan_context);
```

Comment:
For some reason, in all uses of the plan_tree_walker function, the same context is passed to it as it comes to the called function. While you have some new context allocated on the stack.

Reply:
Without `if (IsA(node, Subplan)` clause the `plan_tree_walker` have been failing with [Assertion ](https://github.com/arenadata/gpdb/blob/6ed03a1d78ef8ab6560dc9a5b40a47cd75faaa75/src/backend/optimizer/util/walkers.c#L441) error (if we pass NULL as context), or with SIGISEGV without appropriate `SubplanWalkerContext` [here](https://github.com/arenadata/gpdb/blob/6ed03a1d78ef8ab6560dc9a5b40a47cd75faaa75/src/backend/optimizer/util/walkers.c#L562). Probably I need to find another solution without using walkers?

Reply:
No, why, probably, plan_tree_walker should be used. Only you do it in a strange way.

Reply:
The problem is that without passing appropriate context to `plan_tree_walker` it can fail during processing Subplan node

Reply:
I think, you should typedef struct with plan_tree_base_prefix base, and then init it by planner_init_plan_tree_base, and then call contain_modifyTable_walker

Reply:
see for example motion_sanity_check

Reply:
done